File: postfix-calculator/src/Stack.js
/** Class representing the stack data structure, therefore, with its main property as bein a LIFO (Last In First Out). */

export class Stack {
    constructor() {
        /**
        * Create an empty Array with one property called items (hence to check the items from the array, must do: 
        * object.items). An Array is a native JavaScript object, where they are mutable, pass-by-reference (the
        * objects are not duplicated, both references the same memory location), dynamic (grow and reduce according
        * to needs), and heterogeneous (can store any data type simultaneously)
        */
        this.items = new Array();
    }

    /**
     * Push an item to the array.
     * @return {void}
     * @example
     * stack.push(3); // items is now [3]
     */
    push(value) {
        this.items.push(value);
    }

    /**
     * Pop an item from the array.
     * @return {T | undefined}
     * @example
     * stack.push(3);
     * stack.pop(); // returns 3, items is now []
     */
    pop() {
        return this.items.pop();
    }

    /**
     * View the last item from the array without removing it, representig the top of the stack.
     * @returns {T | undefined}
     * @example
     * stack.push(3);
     * stack.peek(); // returns 3, items is still [3]
     */
    peek() {
        return this.items[this.items.length - 1];
    }

    /**
     * Check if the stack is empty.
     * @return {boolean}
     * @example
     * stack.isEmpty(); // returns true
     */
    isEmpty() {
        return this.items.length === 0;
    }

    /**
     * Prints the stack.
     * @return {array}
     * @example
     * stack.push(3);
     * stack.printStack(); // returns [3]
     */
    printStack() {
        return this.items;
    }
}

File: Data Structures/stack.py
class Node:
    def __init__(self, data, next=None):
        self.data = data
        self.next = next


class Stack:
    def __init__(self):
        self.top = None

    def push(self, data):
        self.top = Node(data, self.top)

    def pop(self):
        if self.is_empty():
            raise IndexError("pop from empty stack")
        data = self.top.data
        self.top = self.top.next
        return data

    def peek(self):
        if self.is_empty():
            raise IndexError("peek from empty stack")
        return self.top.data

    def is_empty(self):
        return self.top is None

    def __repr__(self):
        items = []
        curr = self.top
        while curr:
            items.append(str(curr.data))
            curr = curr.next
        return " -> ".join(items) + " -> None"

s = Stack()
s.push(1)
s.push(2)
s.push(3)
print(s)        # 3 -> 2 -> 1 -> None
print(s.peek()) # 3
print(s.pop())  # 3
print(s.pop())  # 2
print(s)        # 1 -> None
print(s.is_empty())  # False
s.pop()
print(s.is_empty())  # True